rx-java/src/main/java/io/vertx/rx/java/ReadStreamSubscriber.java

```diff
@@ -0,0 +1,182 @@
+
+  public static final int BUFFER_SIZE = 16;
+
+  public static <R, J> ReadStream<J> asReadStream(Observable<R> observable, Function<R, J> adapter) {
+    ReadStreamSubscriber<R, J> observer = new ReadStreamSubscriber<>(adapter);
+    observable.subscribe(observer);
+    return observer;
+  }
+
+  private final Function<R, J> adapter;
+  private Handler<Void> endHandler;
+  private Handler<Throwable> exceptionHandler;
+  private Handler<J> elementHandler;
+  private boolean paused = false;
+  private Throwable completed;
+  private ArrayDeque<R> pending = new ArrayDeque<>();
+  private int requested = 0;
+
+  public ReadStreamSubscriber(Function<R, J> adapter) {
+    this.adapter = adapter;
+    request(0);
+  }
+
+  @Override
+  public ReadStream<J> handler(Handler<J> handler) {
+    synchronized (this) {
+      elementHandler = handler;
+    }
+    checkStatus();
+    return this;
+  }
+
+  @Override
+  public ReadStream<J> pause() {
+    synchronized (this) {
+      paused = true;
+    }
+    return this;
+  }
+
+  @Override
+  public ReadStream<J> resume() {
+    synchronized (this) {
+      paused = false;
+    }
+    checkStatus();
+    return this;
+  }
+
+  private void checkStatus() {
+    Runnable action = NOOP_ACTION;
+    while (true) {
+      J adapted;
+      Handler<J> handler;
+      synchronized (this) {
+        if (!paused && (handler = elementHandler) != null && pending.size() > 0) {
+          requested--;
+          R item = pending.poll();
+          adapted = adapter.apply(item);
+        } else {
+          if (completed != null) {
+            if (pending.isEmpty()) {
+              Handler<Throwable> onError;
+              Throwable result;
+              if (completed != DONE_SENTINEL) {
+                onError = exceptionHandler;
+                result = completed;
+                exceptionHandler = null;
+              } else {
+                onError = null;
+                result = null;
+              }
+              Handler<Void> onCompleted = endHandler;
+              endHandler = null;
+              action = () -> {
+                try {
+                  if (onError != null) {
+                    onError.handle(result);
+                  }
+                } finally {
+                  if (onCompleted != null) {
+                    onCompleted.handle(null);
+                  }
+                }
+              };
+            }
+          } else if (requested < BUFFER_SIZE / 2) {
+            int request = BUFFER_SIZE - requested;
+            action = () -> request(request);
+            requested = BUFFER_SIZE;
+          }
+          break;
+        }
+      }
+      handler.handle(adapted);
+    }
+    action.run();
+  }
+
+  @Override
+  public ReadStream<J> endHandler(Handler<Void> handler) {
+    synchronized (this) {
+      if (completed == null || pending.size() > 0) {
+        endHandler = handler;
+      } else {
+        if (handler != null) {
```

Comment:
I guess the logic is that null is a valid argument even when the stream is completed and there are no pending buffers.